File: src/ai_multi_reference_timekeeping/references.py
"""Reference source simulators for testing and experimentation.

These classes provide a controlled way to generate offset measurements with
known noise characteristics, which is useful for unit tests and notebook demos.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable

import random

from .fusion import Measurement


@dataclass
class ReferenceSource:
    """A simulated timing reference that produces noisy offset measurements.

    Attributes:
        name: Identifier for the reference source.
        noise_std: Standard deviation of measurement noise in seconds.
        bias: Optional constant bias added to measurements.
        drift_bias: Optional drift bias applied over time (seconds/second).
    """

    name: str
    noise_std: float
    bias: float = 0.0
    drift_bias: float = 0.0

    def measure(self, true_offset: float, elapsed: float) -> Measurement:
        """Generate a noisy measurement for the given true offset.

        Args:
            true_offset: The simulated true offset between the clock and reference.
            elapsed: Elapsed time in seconds (used for drift bias).

        Returns:
            A Measurement including the noise variance.
        """

        if self.noise_std <= 0:
            raise ValueError("noise_std must be positive")

        noise = random.gauss(0.0, self.noise_std)
        biased_offset = true_offset + self.bias + self.drift_bias * elapsed + noise
        return Measurement(
            name=self.name,
            offset=biased_offset,
            variance=self.noise_std**2,
        )


class ReferenceEnsemble:
    """Collection of reference sources that generate measurements together."""

    def __init__(self, *sources: ReferenceSource) -> None:
        if not sources:
            raise ValueError("At least one reference source is required")
        self._sources = sources

    def snapshot(self, true_offset: float, elapsed: float) -> list[Measurement]:
        """Generate measurements from all sources for the current step."""

        return [source.measure(true_offset, elapsed) for source in self._sources]


class DeterministicReference:
    """Reference source driven by a deterministic callback.

    The deterministic path is useful for tests, because it avoids reliance on
    random number generation while still exercising the fusion logic.
    """

    def __init__(self, name: str, variance: float, generator: Callable[[float], float]) -> None:
        if variance <= 0:
            raise ValueError("variance must be positive")
        self._name = name
        self._variance = variance
        self._generator = generator

    def measure(self, elapsed: float) -> Measurement:
        """Generate a deterministic measurement for a given elapsed time."""

        return Measurement(
            name=self._name,
            offset=self._generator(elapsed),
            variance=self._variance,
        )

File: cpp/src/references.cpp
#include "aimrt/references.hpp"

#include <stdexcept>

namespace aimrt {

Measurement ReferenceSource::measure(double true_offset, double elapsed) const {
    if (noise_std <= 0.0) {
        throw std::runtime_error("noise_std must be positive");
    }
    static thread_local std::mt19937 rng{std::random_device{}()};
    std::normal_distribution<double> dist(0.0, noise_std);
    const double noise = dist(rng);
    const double biased_offset = true_offset + bias + drift_bias * elapsed + noise;
    return Measurement{name, biased_offset, noise_std * noise_std, false, 1.0};
}

ReferenceEnsemble::ReferenceEnsemble(std::vector<ReferenceSource> sources) : sources_(std::move(sources)) {
    if (sources_.empty()) {
        throw std::runtime_error("At least one reference source is required");
    }
}

std::vector<Measurement> ReferenceEnsemble::snapshot(double true_offset, double elapsed) const {
    std::vector<Measurement> output;
    output.reserve(sources_.size());
    for (const auto& source : sources_) {
        output.push_back(source.measure(true_offset, elapsed));
    }
    return output;
}

DeterministicReference::DeterministicReference(std::string name, double variance,
                                               std::function<double(double)> generator)
    : name_(std::move(name)), variance_(variance), generator_(std::move(generator)) {
    if (variance_ <= 0.0) {
        throw std::runtime_error("variance must be positive");
    }
}

Measurement DeterministicReference::measure(double elapsed) const {
    return Measurement{name_, generator_(elapsed), variance_, false, 1.0};
}

}  // namespace aimrt
